src/sql/src/query_model/scalar_expr.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashSet;
+use std::fmt;
+
+use repr::*;
+
+use crate::query_model::{QuantifierId, QuantifierSet};
+
+/// Representation for scalar expressions within a query graph model.
+/// Similar to HirScalarExpr but:
+/// * subqueries are represented as column references to the subquery
+///   quantifiers within the same box the expression belongs to,
+/// * aggregate expressions are considered scalar expressions here
+///   even though they are only valid in the context of a Grouping box,
+/// * column references are represented by a pair (quantifier ID, column
+///   position),
+/// * BaseColumn is used to represent leaf columns, only allowed in
+///   the projection of BaseTables and TableFunctions.
+#[derive(Debug, PartialEq, Clone)]
+pub enum Expr {
+    ColumnReference(ColumnReference),
+    BaseColumn(BaseColumn),
+    Literal(Row, ColumnType),
+}
+
+impl fmt::Display for Expr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self {
+            Expr::ColumnReference(c) => {
+                write!(f, "Q{}.C{}", c.quantifier_id, c.position)
+            }
+            Expr::BaseColumn(c) => {
+                write!(f, "C{}", c.position)
+            }
+            Expr::Literal(row, _) => {
+                write!(f, "{}", row.unpack_first())
```

Comment:
I don't think we can actually use `Literal(Datum, ColumnType)`. The problem is that a `Datum::String` only has a reference to a string instead of the string itself. 
https://github.com/MaterializeInc/materialize/blob/c9080b3065442b5d04dfa2fca9ad07f0067f7ea8/src/repr/src/scalar.rs#L63
Packing a `Datum::String` into a `Row` is the only way to save/move the contents of the string.

Reply:
OK, then a comment next to the `Expr::Literal` variant which indicates that the `Row` field is expected to contain *exactly one* `Datum` field will be helpful.

Reply:
Done.